locale.c

```diff
@@ -2592,46 +2592,68 @@ S_calculate_LC_ALL_string(pTHX_ const char ** category_locales_list,
-     * getting the results of the POSIX 2008 equivalent of
-     *      setlocale(category,  "")
-     * (which gets the locale from the environment).
+     * The return need not be freed by the caller.  This
+     * promise relies on PerlEnv_getenv() returning a mortalized copy to us.
      *
-     * To ensure that we know exactly what those values are, we do the setting
-     * ourselves, using the documented algorithm (assuming the documentation is
-     * correct) rather than use "" as the locale.  This will lead to results
-     * that differ from native behavior if the native behavior differs from the
-     * standard documented value, but khw believes it is better to know what's
-     * going on, even if different from native, than to just guess.
+     * The simulation is needed only on certain platforms; otherwise, libc is
+     * called with "" to get the actual value(s).  The simulation is needed
+     * for:
      *
-     * Another option for the POSIX 2008 case would be, in a critical section,
-     * to save the global locale's current value, and do a straight
-     * setlocale(LC_ALL, "").  That would return our desired values, destroying
-     * the global locale's, which we would then restore.  But that could cause
-     * races with any other thread that is using the global locale and isn't
-     * using the mutex.  And, the only reason someone would have done that is
-     * because they are calling a library function, like in gtk, that calls
-     * setlocale(), and which can't be changed to use the mutex.  That wouldn't
-     * be a problem if this were to be done before any threads had switched,
-     * say during perl construction time.  But this code would still be needed
-     * for the general case.
+     *  1)  On Windows systems, the concept of the POSIX ordering of
+     *      environment variables is missing.  To increase portability of
+     *      programs across platforms, the POSIX ordering is emulated on
+     *      Windows.
+     *
+     *  2)  On POSIX 2008 systems without querylocale(), it is problematic
+     *      getting the results of the POSIX 2008 equivalent of
+     *
+     *          setlocale(category, "")
+     *
+     *      To ensure that we know exactly what those values are, we do the
+     *      setting ourselves, using the documented algorithm specified by the
+     *      POSIX standard (assuming the platform follows the Standard) rather
+     *      than use "" as the locale.  This will lead to results that differ
+     *      from native behavior if the native behavior differs from the
+     *      Standard's documented value, but khw believes it is better to know
+     *      what's going on, even if different from native, than to just guess.
+     *
+     *      Debian systems differ from this standard in having a LANGUAGE
+     *      environment variable used for just LC_MESSAGES.  This function does
+     *      NOT handle that.
```

Comment:
I don't think this is Debian specific, on Fedora 38:
```
[tony@fedora git]$ cat lc_messages.c
#include <stdio.h>
#include <locale.h>
#include <errno.h>

int main() {
  setlocale(LC_ALL, "");
  errno = EINVAL;
  perror("");
}
[tony@fedora git]$ gcc -olc_messages lc_messages.c
[tony@fedora git]$ ./lc_messages 
Invalid argument
[tony@fedora git]$ LANGUAGE=de_DE.utf8 ./lc_messages 
Das Argument ist ungültig
```
Debian does have a patch against glibc to [include the LANGUAGE environment variable in the `locale` tool's output](https://sources.debian.org/patches/glibc/2.36-9%2Bdeb12u1/locale/locale-print-LANGUAGE.diff/) but I don't see anything to indicate Debian are patching this.

The glibc sources indicate this is is a [GNU extension](https://sourceware.org/git/?p=glibc.git;a=blob;f=intl/dcigettext.c;h=7886ac9545e33367eacd8d8ba1497bc45ffbe367;hb=HEAD#l1517), which is presumably why Debian patched `locale` to improve diagnostic reports.

So I'd expect all glibc systems to use `LANGUAGE`.

(This started as "Is this just Debian too, or Ubuntu as well", and down the rabbit hole I went.)

Reply:
I did s/Debian/glibc/ here